src/main/java/com/acrolinx/proxy/AcrolinxProxyHttpServlet.java

```diff
@@ -249,57 +259,48 @@ private String getUsernameFromApplicationSession() {
+  private void modifyRequest(HttpServletRequest httpServletRequest, Builder httpRequestBuilder) {
+    httpRequestBuilder.uri(getTargetUri(httpServletRequest));
+    setRequestHeader(httpRequestBuilder, "User-Agent", "Acrolinx Proxy");
+    setRequestHeader(httpRequestBuilder, "X-Acrolinx-Integration-Proxy-Version", "2");
   }
 
   private void proxyRequest(
       HttpServletRequest httpServletRequest,
       HttpServletResponse httpServletResponse,
-      HttpRequestBase httpRequestBase)
+      Builder httpRequestBuilder)
       throws IOException {
-    copyHeaders(httpServletRequest, httpRequestBase);
+    copyHeaders(httpServletRequest, httpRequestBuilder);
 
-    addAcrolinxUrlHeader(httpServletRequest, httpRequestBase);
+    addAcrolinxBaseUrlHeader(httpServletRequest, httpRequestBuilder);
 
-    modifyRequest(httpServletRequest, httpRequestBase);
+    modifyRequest(httpServletRequest, httpRequestBuilder);
 
-    configureTimeouts(httpRequestBase);
+    httpRequestBuilder.timeout(timeoutDuration);
 
-    // TODO: Make sure not to call the following line in case a user is not
-    // authenticated to the application.
-    addSingleSignOnHeaders(httpRequestBase);
+    addSingleSignOnHeaders(httpRequestBuilder);
 
-    LOGGER.info("Performing HTTP request: {}", httpRequestBase);
+    LOGGER.info("Performing HTTP request: {}", httpRequestBuilder);
 
-    try (CloseableHttpResponse closeableHttpResponse =
-        closeableHttpClient.execute(httpRequestBase)) {
-      int status = closeableHttpResponse.getStatusLine().getStatusCode();
+    try {
+      HttpResponse<InputStream> httpResponse =
+          httpClient.send(
+              httpRequestBuilder.build(), responseInfo -> BodySubscribers.ofInputStream());
+      int status = httpResponse.statusCode();
       LOGGER.debug("Response received: {}", status);
 
       httpServletResponse.setStatus(status);
 
-      transferResponseHeaders(httpServletResponse, closeableHttpResponse.getAllHeaders());
-
-      HttpEntity httpEntity = closeableHttpResponse.getEntity();
+      transferResponseHeaders(httpServletResponse, httpResponse.headers());
 
-      if (httpEntity != null) {
-        transferResponseBodyWitAdditionalHeaders(httpServletResponse, httpEntity);
-      }
-    } catch (ConnectTimeoutException
-        | SocketException
-        | UnknownHostException
-        | SocketTimeoutException e) {
+      transferResponseBodyWitAdditionalHeaders(httpServletResponse, httpResponse);
+    } catch (ConnectException | HttpTimeoutException e) {
       logExceptionAndSendError(httpServletResponse, e, HttpURLConnection.HTTP_BAD_GATEWAY);
-    } catch (SSLHandshakeException | ClientProtocolException e) {
+    } catch (IOException e) {
       logExceptionAndSendError(httpServletResponse, e, HttpURLConnection.HTTP_UNAVAILABLE);
-    } finally {
-      httpRequestBase.releaseConnection();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```

Comment:
In Java, every thread has an _interrupt_ boolean flag. When you `interrupt()` a running thread (= signal the thread to please stop whatever it is doing asap), the _interrupt_ flag of the interrupted thread is set to true and an [InterruptedException](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/InterruptedException.html) is being thrown to unblock the thread from a blocking call (e.g. reading from an InputStream or being inside `Thread.sleep()`).

When you catch an InterruptedException, the interrupt flag is cleared (= set to false again). This is only poorly documented but really important, since `Thread.isInterrupted()` now returns false and one may think that the thread should continue running, instead of stopping asap. Therefore one has to manually set the _interrupt_ flag to true again after catching an InterruptedException. This is what `Thread.currentThread().interrupt()` does.

The "simple" solution would be to not catch InterruptedException in the first place, so that the _interrupt_ flag does not get reset. However, that is not always possible since InterruptedException is a checked exception and not all method signatures allow throwing such a checked exception. This is the case in this class, the `doGet()`, `doPut()`, etc methods in [the parent HttpServlet class](https://tomcat.apache.org/tomcat-11.0-doc/servletapi/jakarta/servlet/http/HttpServlet.html) do not declare an InterruptedException to be thrown, so we have to catch it here and throw a different RuntimeException (I chose IllegalStateException in this particular case) instead.